.changeset/happy-mirrors-stare.md

```diff
@@ -0,0 +1,5 @@
+---
+"@itwin/itwinui-react": minor
+---
+
+Fixed a bug in `ComboBox` where the controlled state (`value` prop) was not given priority over the uncontrolled state. As a result, passing `value={undefined}` now reverts to the uncontrolled state instead of resetting the value. To reset the value, you now need to pass `value={null}`.
```

Comment:
Ohh, right, updated the changeset (https://github.com/iTwin/iTwinUI/pull/2048/commits/2b27f4191c268a109a53d367a0d832b4a46b299e) to mention to use `defaultValue` instead.

I thought of combining the new `defaultValue` prop changeset with this one since this changeset also mentions `defaultValue`. But then I thought against it since I felt that the new `defaultValue` prop might be missed by the users who just scan the first few words/sentences of each bullet point.

Reply:
> * passing `value={undefined}` now reverts to the uncontrolled state instead of resetting the value. To reset the value, you now need to pass `value={null}`.

This part should be undone, as discussed in [above thread](https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1601987520)

Reply:
Yes, handled in https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1605156984.

Reply:
> passing `value={undefined}` or `value={null}` will continue to reset the value. However, `value={null}` will switch to/remain in the *controlled* state. To reset the value and switch to/remain in the *uncontrolled* state, use `value={undefined}`

Is this necessary to mention in the changeset? The `value={undefined}` behavior is the same as before, and `value={null}` already has its own changeset.

Instead of the changeset, it might be useful to mention the explicit `value={undefined}` behavior in the jsdocs.

Reply:
Bumping [this comment](https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1605198373).

Reply:
> `value={null}` already has its own changeset

To confirm, which changeset are you referring to? Because this PR has mentioned `value={null}` only in this changeset. For now, I just rephrased the changeset. (https://github.com/iTwin/iTwinUI/pull/2048/commits/9d2fd03149653ec8fb8c406661713a8ea5f5a1e1).

> Instead of the changeset, it might be useful to mention the explicit `value={undefined}` behavior in the jsdocs.

Explained this in the JSDocs (same commit).

Reply:
Hmm yeah I'm not exactly sure what I was referring to either. 😅 

Just to confirm, are `value={null}` and `value={undefined}` behaving identically now? That's not what I expected. `undefined` means "absence of value" and `null` means "intentionally empty value". So `value={undefined}` should flip to uncontrolled, but `value={null}` should stay controlled.

Reply:
> Just to confirm, are `value={null}` and `value={undefined}` behaving identically now? That's not what I expected. `undefined` means "absence of value" and `null` means "intentionally empty value". So `value={undefined}` should flip to uncontrolled, but `value={null}` should stay controlled.

Yes, that is how it works currently. I added some screen recordings of it from https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1605156984 to the PR description for better visibility.

Reply:
Hmm okay I think I understand now. The behavior is correct, but the confusion was coming from the way the changeset is phrased.

Here's a suggestion where I rewrote the last point to make it more clear:

> Fixed a bug in `ComboBox` where the controlled state (`value` prop) was not given priority over the uncontrolled state.
> 
> As a result:
> * Setting the default value using `value={myDefaultValue}` will no longer work. Instead, use the new `defaultValue` prop.
> * Resetting the value using `value={null}` will now force the ComboBox to remain in *controlled* mode. If you want to reset the value and also switch to *uncontrolled* mode, then use `value={undefined}` instead.

Reply:
Ohh, sounds good, changed to your suggestion with just a small change (https://github.com/iTwin/iTwinUI/pull/2048/commits/d1671a0e1ece526378468cc7415bd39a12f2a626).

```diff
- * Resetting the value using `value={null}` will now force the ComboBox to remain in _controlled_ mode. If you want to reset the value and also switch to _uncontrolled_ mode, then use `value={undefined}` instead.
+ * Resetting the value using `value={null}` will now force the ComboBox to switch to/remain in *controlled* mode. If you want to reset the value and also switch to/remain in *uncontrolled* mode, then use `value={undefined}` instead.
```